runtime/src/accounts_cache.rs

```diff
@@ -0,0 +1,245 @@
+use dashmap::DashMap;
+use solana_sdk::{account::Account, clock::Slot, hash::Hash, pubkey::Pubkey};
+use std::{
+    collections::HashSet,
+    ops::Deref,
+    sync::{
+        atomic::{AtomicBool, AtomicU64, Ordering},
+        Arc, RwLock,
+    },
+};
+
+pub type SlotCache = Arc<SlotCacheInner>;
+
+#[derive(Default, Debug)]
+pub struct SlotCacheInner {
+    cache: DashMap<Pubkey, CachedAccount>,
+    same_account_writes: AtomicU64,
+    same_account_writes_size: AtomicU64,
+    unique_account_writes_size: AtomicU64,
+    is_frozen: AtomicBool,
+}
+
+impl SlotCacheInner {
+    pub fn report_slot_store_metrics(&self) {
+        datapoint_info!(
+            "slot_repeated_writes",
+            (
+                "same_account_writes",
+                self.same_account_writes.load(Ordering::Relaxed),
+                i64
+            ),
+            (
+                "same_account_writes_size",
+                self.same_account_writes_size.load(Ordering::Relaxed),
+                i64
+            ),
+            (
+                "unique_account_writes_size",
+                self.unique_account_writes_size.load(Ordering::Relaxed),
+                i64
+            )
+        );
+    }
+
+    pub fn insert(&self, pubkey: &Pubkey, account: Account, hash: Hash) {
+        if self.cache.contains_key(pubkey) {
+            self.same_account_writes.fetch_add(1, Ordering::Relaxed);
+            self.same_account_writes_size
+                .fetch_add(account.data.len() as u64, Ordering::Relaxed);
+        } else {
+            self.unique_account_writes_size
+                .fetch_add(account.data.len() as u64, Ordering::Relaxed);
+        }
+        self.cache.insert(*pubkey, CachedAccount { account, hash });
+    }
+
+    pub fn get_cloned(&self, pubkey: &Pubkey) -> Option<CachedAccount> {
+        self.cache
+            .get(pubkey)
+            // 1) Maybe can eventually use a Cow to avoid a clone on every read
+            // 2) Popping is only safe if its guaranteed only replay/banking threads
+            // are reading from the AccountsDb
+            .map(|account_ref| account_ref.value().clone())
+    }
+
+    pub fn mark_slot_frozen(&self) {
+        self.is_frozen.store(true, Ordering::SeqCst);
+    }
+
+    pub fn is_frozen(&self) -> bool {
+        self.is_frozen.load(Ordering::SeqCst)
+    }
+}
+
+impl Deref for SlotCacheInner {
+    type Target = DashMap<Pubkey, CachedAccount>;
+    fn deref(&self) -> &Self::Target {
+        &self.cache
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct CachedAccount {
+    pub account: Account,
+    pub hash: Hash,
+}
+
+#[derive(Debug, Default)]
+pub struct AccountsCache {
+    cache: DashMap<Slot, SlotCache>,
+    cached_roots: RwLock<HashSet<Slot>>,
```

Comment:
how about renaming this to `unflushed_roots`?

- prefixing `cached_` here isn't adding any valuable info anymore; this is a field of `AccountsCache` (duh!)
- `un-` prefix indicates something needs to be done.
- This is actually a task queue for `flush`-ing.

Reply:
perfect, done!